tests/unit/test_endpoints_v2.py

```diff
@@ -0,0 +1,555 @@
+                'Region': 'us-east-1',
+                'Bucket': 'arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint',
+                'Endpoint': 'https://mygreatendpoint.com',
+            },
+            RuleSetEndpoint(
+                url='https://mygreatendpoint.com',
+                properties={
+                    "authSchemes": [
+                        {
+                            "name": "v4",
+                            "signingName": "s3",
+                            "signingScope": "us-east-1",
+                            "uriEscapePath": False,
+                        }
+                    ]
+                },
+                headers={
+                    'x-amz-region-set': [
+                        'us-west-2',
+                    ]
+                },
+            ),
+        ),
+        (
+            {
+                'Region': 'us-east-1',
+                'Bucket': 'arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint',
+                'Endpoint': 'http://mygreatendpoint.biz.amazonaws.com',
+            },
+            RuleSetEndpoint(
+                url='mygreatendpoint.biz.amazonaws.com',
+                properties={
+                    "authSchemes": [
+                        {
+                            "name": "v4",
+                            "signingName": "s3",
+                            "signingScope": "us-east-1",
+                            "uriEscapePath": False,
+                        }
+                    ]
+                },
+                headers={
+                    'x-amz-region-set': [
+                        'us-east-1',
+                    ]
+                },
+            ),
+        ),
+    ],
+)
+def test_endpoint_resolution_with_is_valid_host_label(
+    partitions, ruleset_dict, input_parameters, expected_endpoint
+):
+    endpoint_provider = EndpointProvider(ruleset_dict, partitions)
+    endpoint = endpoint_provider.resolve_endpoint(**input_parameters)
+    assert endpoint.url == expected_endpoint.url
+    assert endpoint.properties == expected_endpoint.properties
+    assert endpoint.headers == expected_endpoint.headers
+
+
+def test_endpoint_resolution_error_with_is_valid_host_label(
+    partitions, ruleset_dict
+):
+    params = {
+        'Region': 'us-east-1',
+        'Bucket': 'arn:aws:s3:us-west-2:123456789012:accesspoint:myendpoint',
+        'Endpoint': 'foo://this.is.not:.a.\va!lid.host.label',
+    }
+    endpoint_provider = EndpointProvider(ruleset_dict, partitions)
+    with pytest.raises(EndpointResolutionError) as err:
+        endpoint_provider.resolve_endpoint(**params)
+        assert err.msg == 'Default error message'
+
+
+@pytest.mark.parametrize(
+    "label,allow_subdomains,expected_output",
+    [
+        (None, False, False),
+        (None, True, False),
+        ('https://test.me.please.com', False, False),
+        ('https://s3.us-east-1.amazonaws.com', True, True),
+        ('https://amazon.com', False, True),
```

Comment:
It looks like we're also not testing the case for allow_subdomains but no subdomain existing. That should return `True`.